src/codegen/llvm_typesystem.rs

```diff
@@ -193,13 +197,88 @@ impl<'ctx, 'cast> Castable<'ctx, 'cast> for PointerValue<'ctx> {
+                cast_data.llvm.builder.build_store(struct_ptr, struct_val);
+                struct_ptr.into()
+            }
             _ => unreachable!("Cannot cast pointer value to {}", cast_data.target_type.get_name()),
         }
     }
 }
 
+impl<'ctx, 'cast> Castable<'ctx, 'cast> for ArrayValue<'ctx> {
+    /// Generates a fat pointer struct for an array if the target type is a VLA,
+    /// otherwise returns the value as is.
+    fn cast(self, cast_data: &CastInstructionData<'ctx, 'cast>) -> BasicValueEnum<'ctx> {
+        if !cast_data.target_type.is_vla() {
+            return self.into();
+        }
+        let builder = &cast_data.llvm.builder;
+        let zero = cast_data.llvm.i32_type().const_zero();
+
+        let Ok(associated_type) = cast_data
+            .llvm_type_index
+            .get_associated_type(cast_data.target_type.get_name()) else {
+                unreachable!("Target type of cast instruction does not exist: {}", cast_data.target_type.get_name())
+        };
+
+        // Get array annotation from parent POU and get pointer to array
+        // Get array annotation from parent POU and get pointer to array
+        let Some(StatementAnnotation::Variable { qualified_name, .. }) = cast_data.annotation  else {
+            unreachable!("Undefined reference: {}", cast_data.value_type.get_name())
+        };
+        let array_pointer = cast_data
+            .llvm_type_index
+            .find_loaded_associated_variable_value(qualified_name.as_str())
+            .unwrap_or_else(|| unreachable!("passed array must be in the llvm index"));
+
+        // gep into the original array. the resulting address will be stored in the VLA struct
+        let arr_gep = unsafe { builder.build_in_bounds_gep(array_pointer, &[zero, zero], "outer_arr_gep") };
+
+        // -- Generate struct & arr_ptr --
+        let ty = associated_type.into_struct_type();
+        let vla_struct = builder.build_alloca(ty, "vla_struct");
+
+        let Ok(vla_arr_ptr) = builder.build_struct_gep(vla_struct, 0, "vla_array_gep") else {
+            unreachable!("Must have a valid, GEP-able fat-pointer struct at this stage")
+        };
+
+        let Ok(vla_dimensions_ptr) = builder.build_struct_gep(vla_struct, 1, "vla_dimensions_gep") else {
+            unreachable!("Must have a valid, GEP-able fat-pointer struct at this stage")
+        };
+
+        // -- Generate dimensions --
+        let DataTypeInformation::Array { dimensions, .. } = cast_data.value_type else { unreachable!() };
+        let mut dims = Vec::new();
+        for dim in dimensions {
+            dims.push(dim.start_offset.as_int_value(cast_data.index).unwrap());
+            dims.push(dim.end_offset.as_int_value(cast_data.index).unwrap());
+        }
+
+        // Populate each array element
+        for (i, val) in dims.iter().enumerate() {
+            let value = cast_data.llvm.i32_type().const_int(*val as u64, true);
```

Comment:
I think this could be a bit more concise by constructing an array-value and then storing into the array (since the arrays should not be huge a store should be fine)
```st
    let dimensions = dims.iter().map(|it| cast_data.llvm.i32_type().const_int(*it as u64, true)).collect::<Vec<_>>();
    let array_value = cast_data.llvm.i32_type().const_array(&dimensions);
    builder.build_store(vla_dimensions_ptr, array_value);
```